File: src/main/java/ch/uzh/ifi/access/repository/TaskFileRepository.java
package ch.uzh.ifi.access.repository;

import ch.uzh.ifi.access.model.TaskFile;
import jakarta.transaction.Transactional;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.security.access.prepost.PostFilter;

import java.util.List;
import java.util.Optional;

public interface TaskFileRepository extends JpaRepository<TaskFile, Long> {

    @Transactional
    @PostFilter("hasRole(filterObject.task.assignment.course.slug + '-assistant') or filterObject.published")
    List<TaskFile> findByTask_IdAndEnabledTrueOrderByIdAscPathAsc(Long taskId);

    @Transactional
    List<TaskFile> findByTask_IdAndEnabledTrue(Long taskId);

    Optional<TaskFile> findByTask_IdAndPath(Long taskId, String filePath);
}

File: src/main/kotlin/ch/uzh/ifi/access/repository/TaskFileRepository.kt
package ch.uzh.ifi.access.repository

import ch.uzh.ifi.access.model.ProblemFile
import jakarta.transaction.Transactional
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.security.access.prepost.PostFilter

interface TaskFileRepository : JpaRepository<ProblemFile?, Long?> {
    @Transactional
    @PostFilter("hasRole(filterObject.problem.assignment.course.slug + '-assistant') or filterObject.isPublished")
    fun findByProblem_IdAndEnabledTrueOrderByIdAscPathAsc(problemId: Long?): List<ProblemFile>

    @Transactional
    fun findByProblem_IdAndEnabledTrue(problemId: Long?): List<ProblemFile>
    fun findByProblem_IdAndPath(taskId: Long?, filePath: String?): ProblemFile?
}